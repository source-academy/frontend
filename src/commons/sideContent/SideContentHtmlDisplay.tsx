import React, { useEffect } from 'react';

type SideContentHtmlDisplayProps = {
  content: string;
  handleAddHtmlConsoleError: (errorMsg: string) => void;
};

const ERROR_MESSAGE_REGEX = /^Line \d+: /i;

const SideContentHtmlDisplay: React.FC<SideContentHtmlDisplayProps> = props => {
  const { content, handleAddHtmlConsoleError } = props;

  useEffect(() => {
    const handleEvent = (event: MessageEvent) => {
      const msg = event.data;
      // Only displays message if it matches the error message format,
      // since there may be other message events not sent by iframe
      if (typeof msg === 'string' && msg.match(ERROR_MESSAGE_REGEX)) {
        handleAddHtmlConsoleError(msg);
      }
    };

    window.addEventListener('message', handleEvent);

    return () => window.removeEventListener('message', handleEvent);
  });
  return (
    <iframe
      className="sa-html-display"
      title="HTML Display"
      sandbox="allow-scripts"
      srcDoc={content}
      src="about:blank"
    />
  );
};

export default SideContentHtmlDisplay;
