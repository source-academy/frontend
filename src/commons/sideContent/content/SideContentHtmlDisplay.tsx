import { IconNames } from '@blueprintjs/icons';
import React, { useEffect } from 'react';
import { connect, MapDispatchToProps } from 'react-redux';
import { bindActionCreators } from 'redux';
import { ResultOutput } from 'src/commons/application/ApplicationTypes';

import { beginAlertSideContent } from '../SideContentActions';
import { SideContentLocation, SideContentTab, SideContentType } from '../SideContentTypes';

type OwnProps = {
  content: string;
  handleAddHtmlConsoleError: (errorMsg: string) => void;
  workspaceLocation: SideContentLocation;
};

type DispatchProps = {
  alertSideContent: () => void;
};

const ERROR_MESSAGE_REGEX = /^Line \d+: /i;

const SideContentHtmlDisplayBase: React.FC<OwnProps & DispatchProps> = props => {
  const { content, handleAddHtmlConsoleError, alertSideContent } = props;

  useEffect(() => {
    const handleEvent = (event: MessageEvent) => {
      const msg = event.data;
      // Only displays message if it matches the error message format,
      // since there may be other message events not sent by iframe
      if (typeof msg === 'string' && msg.match(ERROR_MESSAGE_REGEX)) {
        handleAddHtmlConsoleError(msg);
      }
    };

    window.addEventListener('message', handleEvent);

    return () => window.removeEventListener('message', handleEvent);
  });

  useEffect(() => {
    alertSideContent();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <iframe
      className="sa-html-display"
      title="HTML Display"
      sandbox="allow-scripts"
      srcDoc={JSON.parse(content)}
      src="about:blank"
    />
  );
};

const mapDispatchToProps: MapDispatchToProps<DispatchProps, OwnProps> = (dispatch, props) =>
  bindActionCreators(
    {
      alertSideContent: () =>
        beginAlertSideContent(SideContentType.htmlDisplay, props.workspaceLocation)
    },
    dispatch
  );

export const SideContentHtmlDisplay = connect(null, mapDispatchToProps)(SideContentHtmlDisplayBase);

const makeHtmlDisplayTabFrom = (
  output: ResultOutput,
  handleError: (errorMsg: string) => void,
  workspaceLocation: SideContentLocation
): SideContentTab => ({
  label: 'HTML Display',
  iconName: IconNames.MODAL,
  body: (
    <SideContentHtmlDisplay
      workspaceLocation={workspaceLocation}
      content={output.value}
      handleAddHtmlConsoleError={handleError}
    />
  ),
  id: SideContentType.htmlDisplay
});
export { makeHtmlDisplayTabFrom as default, type OwnProps as SideContentHtmlDisplayProps };
