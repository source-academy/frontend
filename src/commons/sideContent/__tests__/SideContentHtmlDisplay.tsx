import { fireEvent } from '@testing-library/react';
import { mount } from 'enzyme';

import SideContentHtmlDisplay from '../SideContentHtmlDisplay';

test('HTML Display renders correctly', () => {
  const mockProps = {
    content: '<p>Hello World!</p>',
    handleAddHtmlConsoleError: (errorMsg: string) => {}
  };
  const htmlDisplay = mount(<SideContentHtmlDisplay {...mockProps} />);
  expect(htmlDisplay.debug()).toMatchSnapshot();
});

describe('HTML Display postMessage Listener', () => {
  const mockHandleAddHtmlConsoleError = jest.fn((errorMsg: string) => {});

  const mockProps = {
    content: '<p>Hello World!</p>',
    handleAddHtmlConsoleError: mockHandleAddHtmlConsoleError
  };

  beforeAll(() => {
    mount(<SideContentHtmlDisplay {...mockProps} />);
  });

  test('Does not call handleAddHtmlConsoleError if error message format is invalid', async () => {
    const mockMessage = {
      data: 'Invalid error',
      origin: '*'
    };
    fireEvent(window, new MessageEvent('message', mockMessage));
    expect(mockHandleAddHtmlConsoleError).toHaveBeenCalledTimes(0);
  });

  test('Calls handleAddHtmlConsoleError if error message format is valid', async () => {
    const mockMessage = {
      data: 'Line 1: Syntax Error',
      origin: '*'
    };
    fireEvent(window, new MessageEvent('message', mockMessage));
    expect(mockHandleAddHtmlConsoleError).toHaveBeenCalledTimes(1);
    expect(mockHandleAddHtmlConsoleError).toHaveBeenCalledWith(mockMessage.data);
  });
});
