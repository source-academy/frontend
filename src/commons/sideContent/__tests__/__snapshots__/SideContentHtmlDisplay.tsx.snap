// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML Display renders correctly 1`] = `
"<SideContentHtmlDisplay content=\\"<p>Hello World!</p>\\" handleAddHtmlConsoleError={[Function: handleAddHtmlConsoleError]}>
  <iframe className=\\"sa-html-display\\" title=\\"HTML Display\\" sandbox=\\"allow-scripts\\" srcDoc=\\"<p>Hello World!</p>\\" src=\\"about:blank\\" />
</SideContentHtmlDisplay>"
`;
